### PR TITLE
chore: Refactor SDK structure

### LINKED ISSUE #437: [refactor] shared/types is too big, split it
shared/types has gotten pretty big, and we already have other types in other files (events).

We should create a `/types` dir with an index file that exports a number of other types files.

see: https://github.com/open-feature/js-sdk/pull/436#discussion_r1205824805

### PR DESCRIPTION
**Pull request by lukas-reining**
## This PR
Splits up the types file and restructures the SDK by functionalities. Feeling quite good with the structure now.
The PR became quite huge, but I could even move one or two more things (for the client) to the shared implementation.
@toddbaert @beeme1mr Please give feedback if you think the structure is completely off or something seems odd to you🙂

### Related Issues

Fixes #437

### CODE REVIEW

**Comment by toddbaert** [comment]
Will take a look at this tomorrow at the latest (catching up from some time away)

**Comment by toddbaert** [comment]
> but I could even move one or two more things (for the client) to the shared implementation

Feel free to do so in a new commit on this same branch, and I will review it.

As long as there are no functional changes and all changes are just refactors, I'm fine with any additional changes. I don't like mixing functional changes and refactors, but if it's pure refactor, I'm OK with putting as much as you'd like in here.

File: packages/client/src/open-feature.ts [hunk 1]
```diff
@@ -36,31 +34,16 @@ export class OpenFeatureAPI extends OpenFeatureCommonAPI<Provider> {
...
   async setContext(context: EvaluationContext): Promise<void> {
     const oldContext = this._context;
     this._context = context;
     await this._defaultProvider?.onContextChange?.(oldContext, context);
   }
 
+  getContext(): EvaluationContext {
+    return this._context;
+  }
+
```
Comment (toddbaert) [inline packages/client/src/open-feature.ts:46]: Same as [here](https://github.com/open-feature/js-sdk/pull/473/files#r1253316673)?

**Comment by lukas-reining** [comment]
> > but I could even move one or two more things (for the client) to the shared implementation
> 
> Feel free to do so in a new commit on this same branch, and I will review it.
> 
> As long as there are no functional changes and all changes are just refactors, I'm fine with any additional changes. I don't like mixing functional changes and refactors, but if it's pure refactor, I'm OK with putting as much as you'd like in here.

Oh I meant I already was able and did so.
So I moved as much as I think I could already 🙂

**Comment by toddbaert** [comment]
> > > but I could even move one or two more things (for the client) to the shared implementation
> > 
> > 
> > Feel free to do so in a new commit on this same branch, and I will review it.
> > As long as there are no functional changes and all changes are just refactors, I'm fine with any additional changes. I don't like mixing functional changes and refactors, but if it's pure refactor, I'm OK with putting as much as you'd like in here.
> 
> Oh I meant I already was able and did so. So I moved as much as I think I could already slightly_smiling_face

So far it looks beautiful.

**Comment by toddbaert** [comment]
Besides the too small findings above (`getContext`) this looks good to me. I'm going to do some testing locally to make sure we don't have any weird packaging issues or import cycles. As long as I don't find something like that I'll approve.

File: packages/server/src/open-feature.ts [hunk 2]
```diff
@@ -38,25 +40,15 @@ export class OpenFeatureAPI extends OpenFeatureCommonAPI<Provider> {
...
-
   setContext(context: EvaluationContext): this {
     this._context = context;
     return this;
   }
 
+  getContext(): EvaluationContext {
+    return this._context;
+  }
+
```
[the remark this answers is not in the document]
Comment (lukas-reining) [reply packages/server/src/open-feature.ts:51]: I think we can not easily move this as it comes from the ManageContext interface.
The getContext could be simply moved, but not the setContext as this returns something different in both SDKs.

For the Server we use `class OpenFeatureAPI extends OpenFeatureCommonAPI<Provider> implements ManageContext<OpenFeatureAPI> ` which results in `setContext(context: EvaluationContext): this`

For the client we use `class OpenFeatureAPI extends OpenFeatureCommonAPI<Provider> implements ManageContext<Promise<void>>` which results in `async setContext(context: EvaluationContext): Promise<void>`.

My proposal would be that we change the client SDK, to also return the OpenFeatureApi then we could move this to the common API. If we do not want to change the return type in the SDK, I think we can not move this up.

[the remark this answers is not in the document]
Comment (toddbaert) [reply packages/server/src/open-feature.ts:51]: I dont think we want to change that rerturn type, it will break some chaining, I think.

**Review by toddbaert — APPROVED** [review]
I checked this locally, packaged for ESM, CJS and web, and it worked in 3 different consumer projects. The structure is very easy to follow.I checked this locally, packaged for ESM, CJS and web, and it worked in 3 different consumer projects. The structure is very easy to follow.

**Pull request merged**